### ISSUE TITLE
Rake not working during package installation

### ISSUE THREAD

**Issue by mdorier** [LLNL/spack#24731]
I'm trying to make a spack package for [mruby](http://mruby.org), an implementation of Ruby for embedded systems. This package depends on `ruby` and on `rake`. I can easily build it manually as follows (provided the ruby and ruby-rake packages are installed and loaded):

```
git clone https://github.com/mruby/mruby.git
cd mruby
rake
```

(it doesn't have a `rake install` target so the next step if I want the headers and libs in another location is to copy them manually).

I have tried to reproduce this build process as a spack package as follows:

```
from spack import *


class Mruby(RubyPackage):
    """mruby is the lightweight implementation of the Ruby language complying
    to (part of) the ISO standard. Its syntax is Ruby 2.x compatible."""

    homepage = "https://mruby.org/"
    url      = "https://github.com/mruby/mruby/archive/refs/tags/3.0.0.tar.gz"

    version('3.0.0', sha256='95b798cdd931ef29d388e2b0b267cba4dc469e8722c37d4ef8ee5248bc9075b0')

    depends_on('ruby', type=('build'))
    depends_on('ruby-rake', type=('build'))
    depends_on('bison', type=('build'))

    def install(self, spec, prefix):
        rake()
        # mkdir/install files here ...
```

However I'm getting this error from the `rake` step:

```
==> Error: ProcessError: Command exited with status 1:
    '/projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/bin/rake'

1 error found in build log:
     6       Successfully built RubyGem
     7       Name: mruby-source
     8       Version: 3.0.0
     9       File: mruby-source-3.0.0.gem
     10    ==> mruby: Executing phase: 'install'
     11    ==> [2021-07-06-18:13:15.730770] '/projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4n
           jvp3pyba2f42k/bin/rake'
  >> 12    /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/lib/ruby/3.0.0/rubyg
           ems.rb:281:in `find_spec_for_exe': can't find gem rake (>= 0.a) with executable rake (Gem::GemNotFoundException)
     13    	from /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/lib/ruby/3.0.0
           /rubygems.rb:300:in `activate_bin_path'
     14    	from /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/bin/rake:23:in
            `<main>'
```

(ignore lines 6 to 9, they stem from the fact that the package inherits from RubyPackage, but the source is not actually a gem)

The `rake` binary invoked is actually a small ruby script in ruby's installation directory, which loads the rake gem. It seems it is unable to do so within the context of this install function.

Again, the whole process works fine when I build manually. That said, I'm not sure that ruby's `rake` script actually uses the spack-install `ruby-rake` package; it might very well be using the system's.

How can I make ruby correctly find the rake gem in this context?

There must be something related to the `GEM_HOME` or `GEM_PATH` environment variables, but I'm not familiar enough with Ruby to know how to set those variables. The value of `GEM_PATH` in the install function is as follows:

```
/projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/mruby-3.0.0-zqrfyx42dplkmnimvbahfhmdu6hgobek:/projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-rake-13.0.1-2pf7szotsnolhiwv45jwi6bu3uqghfyh
```
so it seems to already correctly have the path to ruby-rake.

**Comment by mdorier** [comment]
I can confirm that when I build manually, when I spack-load `ruby` and `ruby-rake`, despite the `rake` executable being the one from the spack-installed ruby, the gem it uses is the one installed on the system, not the one installed by spack. This is a problem by itself. Here is what `gem environment` gives me:

```
$ gem environment
RubyGems Environment:
  - RUBYGEMS VERSION: 3.2.3
  - RUBY VERSION: 3.0.0 (2020-12-25 patchlevel 0) [x86_64-linux]
  - INSTALLATION DIRECTORY: /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/lib/ruby/gems/3.0.0
  - USER INSTALLATION DIRECTORY: /home/mdorier/.local/share/gem/ruby/3.0.0
  - RUBY EXECUTABLE: /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/bin/ruby
  - GIT EXECUTABLE: /usr/bin/git
  - EXECUTABLE DIRECTORY: /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/bin
  - SPEC CACHE DIRECTORY: /home/mdorier/.local/share/gem/specs
  - SYSTEM CONFIGURATION DIRECTORY: /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/etc
  - RUBYGEMS PLATFORMS:
     - ruby
     - x86_64-linux
  - GEM PATHS:
     - /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/lib/ruby/gems/3.0.0
  - GEM CONFIGURATION:
     - :update_sources => true
     - :verbose => true
     - :backtrace => false
     - :bulk_threshold => 1000
  - REMOTE SOURCES:
     - https://rubygems.org/
  - SHELL PATH:
     - /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/bin
     - /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/openssl-1.1.1k-azcpq7vtohziokt4bl35drcoqkboa2ta/bin
     - /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/bin
     - /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/openssl-1.1.1k-azcpq7vtohziokt4bl35drcoqkboa2ta/bin
     - /projects/spack/var/spack/environments/cereal-env/.spack-env/view/bin
     - /projects/spack/bin
     - /usr/local/sbin
     - /usr/local/bin
     - /usr/sbin
     - /usr/bin
     - /sbin
     - /bin
     - /usr/games
     - /usr/local/games
     - /snap/bin
```

The path to `ruby-rake` is not listed as a gem path. `GEM_PATH` and `GEM_HOME` are empty. Even if I set them to the path to `ruby-rake`, `rake` keeps using the system-installed rake gem.

**Comment by mdorier** [comment]
TL;DR: jump to the last 2 paragraphs of this comment to see the solution, and suggestion for modifying the ruby package.

More investigation: the system does not actually have rake (it has ruby, though), so, taking it step by step, from nothing loaded:

```
$ which ruby
/usr/bin/ruby
$ ruby -v
ruby 2.7.2p137 (2020-10-01 revision 5445e04352) [x86_64-linux-gnu]
$ rake -V
Traceback (most recent call last):
	2: from /usr/bin/rake:23:in `<main>'
	1: from /usr/lib/ruby/vendor_ruby/rubygems.rb:300:in `activate_bin_path'
/usr/lib/ruby/vendor_ruby/rubygems.rb:281:in `find_spec_for_exe': can't find gem rake (>= 0.a) with executable rake (Gem::GemNotFoundException)
```

As expected, rake is not found (it's not installed). Let's now load the spack-installed ruby and retry:

```
$ spack load -r ruby
$ which ruby
/projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/bin/ruby
$ ruby -v
ruby 3.0.0p0 (2020-12-25 revision 95aff21468) [x86_64-linux]
$ rake -V
rake, version 13.0.3
```

First strange thing here, ruby finds a rake gem installed somewhere. Strangely enough, spack supposedly install `ruby-rake@13.0.1`, not 13.0.3. Calling `gem environment`, I can see that ruby is looking here for gems:
```
/projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/lib/ruby/gems/3.0.0
```
and when I go into this directory, in the `gems` subdirectory I can find the rake 13.0.3 here. So ruby 3.0.0 seems to be shipping with its own rake gem by default (maybe the ruby-rake package could be updated to not be installable with ruby 3.0.0, or whatever earlier version started to provide it by default).

`GEM_PATH` and `GEM_HOME` are not set, at this point.

Loading spack's `ruby-rake` package does not change the above behavior. Ruby keeps finding its own rake instead of the spack-provided one.

The weird thing is, if ruby has rake by default, why does it fail to use it during the installation process of a package?

If I call `gem environment`, I get the following (eliminating many things):

```
...
  - INSTALLATION DIRECTORY: /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/lib/ruby/gems/3.0.0
...
  - EXECUTABLE DIRECTORY: /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/bin
...
  - GEM PATHS:
     - /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/ruby-3.0.0-6hobkj5ntaqidqjlt4njvp3pyba2f42k/lib/ruby/gems/3.0.0
 ...
```

If I call `gem environment` during the build process of mruby, I get the following:

```
...
  - INSTALLATION DIRECTORY: /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/mruby-3.0.0-zmqdmnmtf3zbjazkb433bzahi7yuo3mt
...
  - EXECUTABLE DIRECTORY: /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/mruby-3.0.0-zmqdmnmtf3zbjazkb433bzahi7yuo3mt/bin
...
  - GEM PATHS:
     - /projects/spack/opt/spack/linux-ubuntu21.04-sandybridge/gcc-10.3.0/mruby-3.0.0-zmqdmnmtf3zbjazkb433bzahi7yuo3mt
...
```

So the problem is there. It seems that spack sets `GEM_HOME` (which is the location where you want gems to be installed), which is then picked up as gem path. The following code, setting `GEM_PATH` before calling `rake()`, makes everything work:

```
rb = spec['ruby']
env['GEM_PATH'] = rb.prefix+'/lib/ruby/gems/'+str(rb.version)
```

**I can see [here in the ruby package](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/ruby/package.py#L89) that `GEM_PATH` is set to all the paths of the ruby packages defined as dependencies. I think it should also add ruby's own gems subdirectory (`/lib/ruby/gems/'+str(rb.version)`) since ruby itself ships with some gems by default.**

Something weird still is that if I make my package depend on `ruby-rake`, then `GEM_PATH` does contain the path to ruby-rake, yet ruby doesn't find the gem. This may hint at either a problem with ruby-rake, or more generally gems installed outside of the ruby tree.
